### ISSUE TITLE
New Rule: no-unused-class-fields

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#18070]
### Rule details

For private class fields, if the field is never referenced as `this.#name` inside of a class method, it will warn.  For public class fields, if the field is never referenced as `this.name` inside of a class method AND the public field doesn't have an initializer, it will warn. (If the field has an initializer, it may be accessed on an instance.)

### Related ECMAScript feature

Class fields and class private fields

### What type of rule is this?

Warns about a potential problem

### Example code

```js
class Example1 {

    #data;

    constructor(data) {
        this.data = data;
    }
}

class Example2 {

    data;

    constructor(data) {
        this.#data = data;
    }
}
```

### Why should this rule be in the core instead of a plugin?

This is a fairly new language feature and corresponds to `no-unused-vars` as a way to prevent potential errors.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by Skalman** [comment]
This is a great idea - it'd be great if ESLint could check for correct usage of fields! I have some thoughts about the private fields.

> For private class fields, if the field is never referenced as `this.#name` inside of a class method, it will warn.

This should probably be stricter. My suggestion:

**For private class fields, if the field is never used  inside of the class, except for its constructor, it should warn.**

Here are some test cases.

```js
// should warn - only used in the constructor (should be replace with a variable)
class Example3 {
  #data;
  constructor() {
    this.#data = 1;
    if (this.#data) {
    }
  }
}

// should warn - not use, only assigned
class Example4 {
  #data = 1;
  method() {
    this.#data = 1;
  }
}

// should not warn - used in the class in a non-method function
class Example5 {
  #data = 1;
  fn = () => this.#data;
}

// should not warn - used in a sub-class constructor
class Example6 {
  #data = 1;
  static SubClass = class {
    constructor(example) {
      if (example.#data) {
      }
    }
  };
}

// should not warn - used in a sub-class constructor
class Example7 {
  #data = 1;
  static method(...examples) {
    return examples.map((x) => x.#data);
  }
}

// should not warn - tested in method
class Example8 {
  #data;
  static method(example) {
    return #data in example;
  }
}
```

**Comment by mdjermanovic** [comment]
For private class fields, we already have rule [no-unused-private-class-members](https://eslint.org/docs/latest/rules/no-unused-private-class-members).

For public class fields, I'm not sure if a field not being referenced inside the class definitely indicates a possible error as it can be assigned and used in a subclass or on an instance.

**Comment by nzakas** [comment]
> For private class fields, we already have rule [no-unused-private-class-members](https://eslint.org/docs/latest/rules/no-unused-private-class-members).

Oh wow. I could have sworn I looked for this first. Ah well. Too many rules to keep track of.

> For public class fields, I'm not sure if a field not being referenced inside the class definitely indicates a possible error as it can be assigned and used in a subclass or on an instance.

I just ran into a bug caused by this, which is why it was top-of-mind. However, I think the private class field use case is much more important, so we can close this. I'll think more about public class fields and what sort of guards I'd like.
